examples/dataflow-custom-templates/cloud_shell_tutorial.md

```diff
@@ -0,0 +1,223 @@
+[Cloud NAT](https://cloud.google.com/nat/docs/overview); per best practice we
+execute the Dataflow job using private IPs
+
+Therefore, this step will:
+
+- Provision a custom network and subnetwork
+- Provision firewall rules
+- Provision a Cloud NAT and its dependent Cloud Router
+
+Run the terraform workflow in
+the [infrastructure/02.network](infrastructure/02.network) directory.
+
+Terraform will ask your permission before provisioning resources.
+If you agree with terraform provisioning resources,
+type `yes` to proceed.
+
+```sh
+DIR=infrastructure/02.network
+terraform -chdir=$DIR init
+terraform -chdir=$DIR apply -var='project=<walkthrough-project-id/>'
+```
+
+## 5. Provision data pipeline IO resources
+
+The Apache Beam example that our Dataflow template executes is a derived word
+count for both [Java](https://github.com/apache/beam/blob/master/examples/java/src/main/java/org/apache/beam/examples/MinimalWordCount.java)
+and [python](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/wordcount_minimal.py).
+
+The word count example requires a source [Google Cloud Storage](https://cloud.google.com/storage) bucket.
+To make the example interesting, we copy all the files from
+`gs://apache-beam-samples/shakespeare/*` to a custom bucket in our project.
+
+Therefore, this step will:
+- Provision a Google Cloud storage bucket
+- Create Google Cloud storage objects to read from in the pipeline
+
+Run the terraform workflow in
+the [infrastructure/03.io](infrastructure/03.io) directory.
+
+Terraform will ask your permission before provisioning resources.
+If you agree with terraform provisioning resources,
+type `yes` to proceed.
+
+```sh
+DIR=infrastructure/03.io
+terraform -chdir=$DIR init
+terraform -chdir=$DIR apply -var='project=<walkthrough-project-id/>'
+```
+
+## 6. Provision the Dataflow template builder
+
+We will use [Cloud Build](https://cloud.google.com/build) to build the
+custom Dataflow template.  There are advantages to using Cloud Build to build
+our custom Dataflow template, instead of performing the necessary commands on
+our local machine.  Cloud Build connects to our version control,
+[GitHub](https://GitHub.com) in this example, so that any changes made to
+a specific branch will automatically trigger a new build of our Dataflow
+template.
+
+Therefore, this step will:
+
+- Provision cloud build trigger that will:
+    1. Run the language specific build process i.e. gradle shadowJar, go build, etc.
+    2. Execute the `gcloud dataflow flex-template` command with relevant arguments.
+
+### 6.1. Fork the repository into your own GitHub organization or personal account
+
+In order to benefit from [Cloud Build](https://cloud.google.com/build), the service
+requires we own this repository; it will not work with a any repository, even
+if it is public.
+
+Therefore, complete these steps before proceeding:
+1) [Fork the repository](https://github.com/GoogleCloudPlatform/professional-services/fork)
```

Comment:
To test this PR, one will need to fork the following repository instead:
https://github.com/damondouglas/professional-services/fork